src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp

```diff
@@ -1690,27 +1690,36 @@ bool C2_MacroAssembler::in_scratch_emit_size() {
   return MacroAssembler::in_scratch_emit_size();
 }
 
-void C2_MacroAssembler::reduce_operation(Register dst, VectorRegister tmp,
-                                         Register src1, VectorRegister src2,
-                                         BasicType bt, REDUCTION_OP op) {
+void C2_MacroAssembler::rvv_reduce_integral(Register dst, VectorRegister tmp,
+                                            Register src1, VectorRegister src2,
+                                            BasicType bt, int opc) {
+  assert(bt == T_BYTE || bt == T_SHORT || bt == T_INT || bt == T_LONG, "unsupported element type");
+
   Assembler::SEW sew = Assembler::elemtype_to_sew(bt);
   vsetvli(t0, x0, sew);
 
   vmv_s_x(tmp, src1);
 
-  switch (op) {
-    case REDUCTION_OP::ADD:
+  switch (opc) {
+    case Op_AddReductionVI:
+    case Op_AddReductionVL:
       vredsum_vs(tmp, src2, tmp);
       break;
-    case REDUCTION_OP::AND:
+    case Op_AndReductionV:
       vredand_vs(tmp, src2, tmp);
       break;
-    case REDUCTION_OP::OR:
+    case Op_OrReductionV:
       vredor_vs(tmp, src2, tmp);
       break;
-    case REDUCTION_OP::XOR:
+    case Op_XorReductionV:
       vredxor_vs(tmp, src2, tmp);
       break;
+    case Op_MaxReductionV:
```

Comment:
We implemented Byte/Short MaxReductionV like this patch before. but there are 2 test cases that failed:
jdk/jdk/incubator/vector/ShortMaxVectorTests.java 
jdk/jdk/incubator/vector/ByteMaxVectorTests.java
The reason is that "make_reduction_input" will return INT_MAX for Byte & Short type.
Have you met this problem?

Reply:
> 

I see that this has been fixed in the previous PR 8271515.
Approved.

Reply:
Hi @yhzhu20, thanks for the review! I think this problem was solved in https://github.com/openjdk/jdk/pull/5873/commits/4e7f606030d67a4f284c375d9c8811f525ae97fd ：
```diff
@@ -1105,7 +1145,9 @@ Node* ReductionNode::make_reduction_input(PhaseGVN& gvn, int opc, BasicType bt)
     case Op_MinReductionV:
       switch (bt) {
         case T_BYTE:
+          return gvn.makecon(TypeInt::make(max_jbyte));
         case T_SHORT:
+          return gvn.makecon(TypeInt::make(max_jshort));
         case T_INT:
           return gvn.makecon(TypeInt::MAX);
         case T_LONG:
@@ -1120,7 +1162,9 @@ Node* ReductionNode::make_reduction_input(PhaseGVN& gvn, int opc, BasicType bt)
     case Op_MaxReductionV:
       switch (bt) {
         case T_BYTE:
+          return gvn.makecon(TypeInt::make(min_jbyte));
         case T_SHORT:
+          return gvn.makecon(TypeInt::make(min_jshort));
         case T_INT:
           return gvn.makecon(TypeInt::MIN);
         case T_LONG:
```
The Function "make_reduction_input" will return different types for Byte & Short than for int now.

Reply:
> Hi @yhzhu20, thanks for the review! I think this problem was solved in [4e7f606](https://github.com/openjdk/jdk/commit/4e7f606030d67a4f284c375d9c8811f525ae97fd) ：
> 
> ```diff
> @@ -1105,7 +1145,9 @@ Node* ReductionNode::make_reduction_input(PhaseGVN& gvn, int opc, BasicType bt)
>      case Op_MinReductionV:
>        switch (bt) {
>          case T_BYTE:
> +          return gvn.makecon(TypeInt::make(max_jbyte));
>          case T_SHORT:
> +          return gvn.makecon(TypeInt::make(max_jshort));
>          case T_INT:
>            return gvn.makecon(TypeInt::MAX);
>          case T_LONG:
> @@ -1120,7 +1162,9 @@ Node* ReductionNode::make_reduction_input(PhaseGVN& gvn, int opc, BasicType bt)
>      case Op_MaxReductionV:
>        switch (bt) {
>          case T_BYTE:
> +          return gvn.makecon(TypeInt::make(min_jbyte));
>          case T_SHORT:
> +          return gvn.makecon(TypeInt::make(min_jshort));
>          case T_INT:
>            return gvn.makecon(TypeInt::MIN);
>          case T_LONG:
> ```
> 
> The function "make_reduction_input" will return different types for Byte & Short than for int now.

Understood. Thank you.

Reply:
> > 
> 
> I see that this has been fixed in the previous PR 8271515. Approved.

Sorry, I didn't refresh the page in time, so I saw your message too late and repeated the same message.